dnsapi/dns_wedos.sh

```diff
@@ -0,0 +1,366 @@
+    _info "WEDOS WAPI: dns record found with id '${_row_id}'"
+
+    if _wapi_delete_row "${_domain}" "${_row_id}"; then
+      _info "WEDOS WAPI: dns row were deleted and changes commited!"
+      return 0
+    fi
+  fi
+
+  _err "Requested dns row were not found or was imposible to delete it, do it manually"
+  _err "Delete: ${fulldomain}"
+  _err "Value: ${txtvalue}"
+  return 1
+}
+
+####################  Private functions below ##################################
+
+# Function _wapi_post(), only takes data, prepares auth token and provide result
+# $1 - WAPI command string, like 'dns-domains-list'
+# $2 - WAPI data for given command, is not required
+# returns WAPI response if request were successfully delivered to WAPI endpoint
+_wapi_post() {
+  command=$1
+  data=$2
+
+  _debug "Command : ${command}"
+  _debug "Data : ${data}"
+
+  if [ -z "${command}" ]; then
+    _err "No command were provided, implamantation error!"
+    return 1
+  fi
+
+  # Prepare authentification token
+  hour=$(date +%H)
+  token=$(printf '%s' "${WEDOS_Authtoken}${hour}" | sha1sum | head -c 40)
+  _debug "Authentification token is '${token}'"
+
+  # Build xml request
+
+  request="request=<?xml version=\"1.0\" encoding=\"UTF-8\"?>\
+<request>\
+  <user>${WEDOS_Username}</user>\
+  <auth>${token}</auth>\
+  <command>${command}</command>"
+
+  if [ -z "${data}" ]; then
+    echo "" 1>/dev/null
+  else
+    request="${request}${data}"
+  fi
+
+  if [ -z "$TESTING_STAGE" ]; then
+    echo "" 1>/dev/null
+  else
+    request="${request}\
+  <test>1</test>"
+  fi
+
+  request="${request}\
+</request>"
+
+  _debug "Request to WAPI is: ${request}"
+
+  if ! response="$(_post "${request}" "$WEDOS_WAPI_ENDPOINT")"; then
+    _err "Error contacting WEDOS WAPI with command ${command}"
+    return 1
+  fi
+
+  _debug "Response : ${response}"
+  echo "${response}" | grep "<code>1000</code>" 1>/dev/null 2>/dev/null
+
+  return "$?"
+}
+
+# _get_root() function, for provided full domain, like _acme_challenge.www.example.com verify if WEDOS contains a primary active domain and found what is subdomain
+# $1 - full domain to verify, ie _acme_challenge.www.example.com
+# build ${_domain} found at WEDOS, like example.com and ${_sub_domain} from provided full domain, like _acme_challenge.www
+_get_root() {
+  domain=$1
+
+  if [ -z "${domain}" ]; then
+    _err "Function _get_root was called without argument, implementation error!"
+    return 1
+  fi
+
+  _debug "Get root for domain: ${domain}"
+
+  _debug "Getting list of domains using WAPI ..."
+
+  if ! _wapi_post "dns-domains-list"; then
+    _err "Error on WAPI request for list of domains, response : ${response}"
+    return 1
+  else
+    _debug "DNS list were successfully retrieved, response : ${response}"
+  fi
+
+  for xml_domain in $(echo "${response}" | tr -d '\012\015' | grep -o -E "<domain>( )*<name>.*</name>( )*<type>primary</type>( )*<status>active</status>" | grep -o -E "<name>.*</name>"); do
```

Comment:
don't use `grep -oE`,  plesae use `_egrep_o()` instead.
It will be even better if you can use just `sed` and `tr` instead.

Reply:
I replaced all the grep -E -o by sed and tr as sugessted. My script is using single grep to get text match in a pipe.